cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -516,6 +601,203 @@ std::shared_ptr<ScalarFunction> MakeUnarySignedArithmeticFunctionNotNull(
+          *output = *array->data();
+          return Status::OK();
+        } else if (arg.is_array() && arg.array()->null_count == length) {
+          *output = *arg.array();
+          return Status::OK();
+        }
+      }
+      // AND together the validity buffers of all arrays
+      ARROW_ASSIGN_OR_RAISE(output->buffers[0], ctx->AllocateBitmap(length));
+      bool first = true;
+      for (const auto& arg : batch.values) {
+        if (!arg.is_array()) continue;
+        auto arr = arg.array();
+        if (!arr->buffers[0]) continue;
+        if (first) {
+          ::arrow::internal::CopyBitmap(arr->buffers[0]->data(), arr->offset, length,
+                                        output->buffers[0]->mutable_data(),
+                                        /*dest_offset=*/0);
+          first = false;
+        } else {
+          ::arrow::internal::BitmapAnd(
+              output->buffers[0]->data(), /*left_offset=*/0, arr->buffers[0]->data(),
+              arr->offset, length, /*out_offset=*/0, output->buffers[0]->mutable_data());
+        }
+      }
+    }
+
+    if (batch.values[0].is_scalar()) {
+      // Promote to output array
+      ARROW_ASSIGN_OR_RAISE(auto array, MakeArrayFromScalar(*batch.values[0].scalar(),
+                                                            length, ctx->memory_pool()));
+      *output = *array->data();
+      if (!batch.values[0].scalar()->is_valid) {
+        // MakeArrayFromScalar reuses the same buffer for null/data in
+        // this case, allocate a real one since we'll be writing to it
+        ARROW_ASSIGN_OR_RAISE(output->buffers[0], ctx->AllocateBitmap(length));
+        ::arrow::internal::BitmapXor(output->buffers[0]->data(), /*left_offset=*/0,
+                                     output->buffers[0]->data(), /*right_offset=*/0,
+                                     length, /*out_offset=*/0,
+                                     output->buffers[0]->mutable_data());
+      }
+    } else {
+      // Copy to output array
+      const ArrayData& input = *batch.values[0].array();
+      ARROW_ASSIGN_OR_RAISE(output->buffers[1], ctx->Allocate(length * sizeof(OutValue)));
+      if (options.skip_nulls && input.buffers[0]) {
+        ARROW_ASSIGN_OR_RAISE(output->buffers[0], ctx->AllocateBitmap(length));
+        ::arrow::internal::CopyBitmap(input.buffers[0]->data(), input.offset, length,
+                                      output->buffers[0]->mutable_data(),
+                                      /*dest_offset=*/0);
+      }
+      // This won't work for nested or variable-sized types
+      std::memcpy(output->buffers[1]->mutable_data(),
+                  input.buffers[1]->data() + (input.offset * sizeof(OutValue)),
+                  length * sizeof(OutValue));
+    }
+
+    for (size_t i = 1; i < batch.values.size(); i++) {
+      OutputArrayWriter<OutType> writer(out->mutable_array());
+      if (batch.values[i].is_scalar()) {
```

Comment:
I think instead it'd be useful to sort the inputs and handle all the scalars first. This would allow you to consider only arrays in this loop